### PR TITLE
[MF-216] Record a Patient Contact Person

### PR DESCRIPTION
**Pull request by marduzca**
image

### CODE REVIEW

File: src/patient-registration/section/contact-person-section.component.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import { TextInput } from '../input//basic-input/text-input.component';
```
Comment (nickjhill14) [inline src/patient-registration/section/contact-person-section.component.tsx]: This import is unused and can be deleted

File: src/patient-registration/patient-registration-validation.test.tsx [hunk 2]
```diff
@@ -42,3 +42,42 @@ describe('telephone number input', () => {
   testInvalidPhoneNumber('+0800001066');
   testInvalidPhoneNumber('(0800)001066');
 });
+
+describe('contact person telephone number input', () => {
```
Comment (nickjhill14) [inline src/patient-registration/patient-registration-validation.test.tsx:46]: On second look, there seems to be a fair amount of duplication with the patient telephone number input. Could we merge the 2 into a single describe (e.g. telephone number fields)?

File: src/patient-registration/patient-registration-helper.tsx [hunk 3]
```diff
@@ -14,6 +14,12 @@ interface AddressValue {
   country: string;
 }
 
+interface ContactPerson {
```
Comment (nickjhill14) [inline src/patient-registration/patient-registration-helper.tsx]: Look good to me 🚀

File: src/patient-registration/input/basic-input/select-input.component.tsx [hunk 4]
```diff
@@ -11,10 +11,10 @@ export const SelectInput: React.FC<SelectInputProps> = ({ name, options }) => {
   const [field, meta] = useField(name);
   const selectOptions = [
     <option key="" value="" disabled>
-      Select {name}
+      Choose
     </option>,
     options.map(currentOption => (
-      <option key={currentOption.charAt(0)} value={currentOption.charAt(0)}>
+      <option key={currentOption} value={currentOption}>
```
Comment (nickjhill14) [inline src/patient-registration/input/basic-input/select-input.component.tsx:17]: This is a nice refactor 🚀

File: src/patient-registration/patient-registration-validation.tsx [hunk 5]
```diff
@@ -4,7 +4,7 @@ export const validationSchema = Yup.object({
   givenName: Yup.string().required('Given name is required'),
   familyName: Yup.string().required('Family name is required'),
   gender: Yup.string()
-    .oneOf(['M', 'F', 'O', 'U'], 'Gender is unspecified')
+    .oneOf(['Male', 'Female', 'Other', 'Unspecified'], 'Gender is unspecified')
```
Comment (nickjhill14) [inline src/patient-registration/patient-registration-validation.tsx]: This may cause issues with form submission. When recording a patient's gender, we only submit the first letter of the word (e.g. 'M' for male)

File: src/patient-registration/patient-registration.component.tsx [hunk 6]
```diff
@@ -31,6 +32,11 @@ export interface FormValues {
   stateProvince: string;
   country: string;
   postalCode: string;
+  contactPersonGivenName: string;
```
Comment (nickjhill14) [inline src/patient-registration/patient-registration.component.tsx:36]: Look good to me. However, we should consider abstracting some of these values into smaller interfaces as this interface is too large.

File: src/patient-registration/patient-registration.component.tsx [hunk 7]
```diff
@@ -115,6 +126,21 @@ export const PatientRegistration: React.FC = () => {
       },
     };
 
+    if (values.contactPersonGivenName || values.contactPersonMiddleName || values.contactPersonFamilyName) {
```
Comment (nickjhill14) [inline src/patient-registration/patient-registration.component.tsx]: If a contact person is being recorded, then we need their given name and family name to be recorded. We should update the acceptance criteria to reflect that and we could use Formik/Yup validation to check that those fields are not empty.

File: src/patient-registration/patient-registration.component.tsx [hunk 8]
```diff
@@ -135,7 +161,8 @@ export const PatientRegistration: React.FC = () => {
             <h1 className={`omrs-type-title-1 ${styles.title}`}>New Patient</h1>
             <DemographicsSection setFieldValue={props.setFieldValue} values={props.values} />
             <ContactInfoSection />
-            <button className={`omrs-btn omrs-filled-action ${styles.submit}`} type="submit">
+            <ContactPersonSection />
+            <button className={`omrs-btn omrs-filled-action ${styles.submit}`} type="submit" name="registerButton">
```
Comment (nickjhill14) [inline src/patient-registration/patient-registration.component.tsx]: Why have we added a name attribute here? For accessibility reasons?

File: src/patient-registration/patient-registration.test.tsx [hunk 9]
```diff
@@ -33,7 +34,7 @@ describe('demographics section', () => {
   it('updates to correct gender', async () => {
     const { container } = render(<PatientRegistration />);
     const genderSelect = container.querySelector('select[name="gender"]') as HTMLSelectElement;
-    const expectedGender = 'M';
+    const expectedGender = 'Male';
```
Comment (nickjhill14) [inline src/patient-registration/patient-registration.test.tsx:36]: See previous comments regarding this. I believe this is also out of scope for this ticket.

File: src/patient-registration/patient-registration.test.tsx [hunk 10]
```diff
@@ -3,6 +3,7 @@ import ReactDOM from 'react-dom';
 import { render, fireEvent, wait, getByLabelText } from '@testing-library/react';
 import dayjs from 'dayjs';
 import { PatientRegistration } from './patient-registration.component';
+import { act } from 'react-dom/test-utils';
```
Comment (nickjhill14) [inline src/patient-registration/patient-registration.test.tsx]: This is an unused import and can be deleted.

File: src/patient-registration/patient-registration.test.tsx [hunk 11]
```diff
@@ -77,3 +78,97 @@ describe('contact info section', () => {
   updateAddress('country');
   updateAddress('postalCode');
 });
+
+describe('contact person section', () => {
```
Comment (nickjhill14) [inline src/patient-registration/patient-registration.test.tsx:81]: Looks good to me. We may need another ticket to ensure that all our section/input tests are consistent in the way that the tests are setup and structured.

File: src/patient-registration/section/contact-person-section.component.tsx [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+import { SelectInput } from '../input//basic-input/select-input.component';
+import { NameInput } from '../input/custom-input/name-input.component';
+import styles from './section.css';
+import { TelephoneNumberInput } from '../input/basic-input/telephone-number-input/telephone-number-input.component';
+
+export const ContactPersonSection: React.FC = () => {
```
Comment (nickjhill14) [inline src/patient-registration/section/contact-person-section.component.tsx:7]: Looks good to me 🚀

**Review by nickjhill14 — COMMENTED** [review]
I'm assuming that the form submission is throwing an error response? I can't currently test it as the server is down. If that is the case though, we can pair to get that set up through making new person attributes.

Another thing to note is that because this feature is so large, I think we should merge it into `openmrs-patient-registration:patient-contact-person` rather than the `openmrs-patient-registration:develop` for the time being. We may need to implement a feature toggle for this feature as well, but we can discuss that further.

File: src/patient-registration/patient-registration-validation.test.tsx (hunk 2 above)
Comment (PB-TW) [reply src/patient-registration/patient-registration-validation.test.tsx:46]: Think we should do that as part of the validation ticket. That involves quite some refactoring anyway. I feel it makes sense like this as part of the contact person context.

Comment (nickjhill14) [reply src/patient-registration/patient-registration-validation.test.tsx:46]: Good point, I'll mark this as resolved

File: src/patient-registration/patient-registration.component.tsx (hunk 6 above)
Comment (marduzca) [reply src/patient-registration/patient-registration.component.tsx:36]: We started working on this but it seems more complicated than expected, so maybe it would make sense to make this a story?

File: src/patient-registration/patient-registration.test.tsx (hunk 9 above)
Comment (marduzca) [reply src/patient-registration/patient-registration.test.tsx:36]: We kept the selectInput general and for the gender we make sure that the payload contains only the first letter.

File: src/patient-registration/patient-registration.component.tsx (hunk 6 above)
Comment (nickjhill14) [reply src/patient-registration/patient-registration.component.tsx:36]: Yeah, I think it would be in a different story as it's changes would probably be applied to other form fields as well.

File: src/patient-registration/patient-registration-validation.tsx [hunk 13]
```diff
@@ -14,4 +14,16 @@ export const validationSchema = Yup.object({
   monthsEstimated: Yup.number().min(0, 'Months cannot be less than 0'),
   telephoneNumber: Yup.string().matches(/^[0-9]*$/, 'Telephone number should only contain digits'),
   contactPersonPhone: Yup.string().matches(/^[0-9]*$/, 'Telephone number should only contain digits'),
+  contactPersonGivenName: Yup.string().when(['contactPersonPhone', 'contactPersonRelationship'], {
```
Comment (nickjhill14) [inline src/patient-registration/patient-registration-validation.tsx]: This is a nice implementation for handling the validation` onSubmit`.
However, I've noticed 2 issues:
1. If the user inputs a middle name and nothing else, no error message is shown for given and family names being required. `contactPersonMiddleName` could be added to the `when` arguments.
2. No error message/validation is shown until the `Register Patient` button is clicked. We would want to inform the user as soon as they have 'touched' or inputted any data into` contactPersonPhone`, `contactPersonRelationship`, and `contactPersonMiddleName`. This could be achieved with `onBlur` or `touched` I think.

File: src/patient-registration/patient-registration-validation.tsx [hunk 14]
```diff
@@ -14,4 +14,16 @@ export const validationSchema = Yup.object({
...
+    is: (contactPersonPhone, contactPersonRelationship) =>
+      (contactPersonPhone && contactPersonPhone.length > 0) ||
+      (contactPersonRelationship && contactPersonRelationship.length > 0),
+    then: Yup.string().required('Given name is required'),
+  }),
+  contactPersonFamilyName: Yup.string().when(['contactPersonPhone', 'contactPersonRelationship'], {
+    is: (contactPersonPhone, contactPersonRelationship) =>
```
Comment (nickjhill14) [inline src/patient-registration/patient-registration-validation.tsx]: Is it possible to extract these two functions into a single one to remove some duplication?

**Review by nickjhill14 — CHANGES_REQUESTED** [review]
I've added some comments in regards to the Yup validation. It looks great but can be tweaked a little 🚀

**Review by nickjhill14 — APPROVED** [review]

**Comment by nickjhill14** [comment]
The last things to do for this ticket are:

- [x] Adding a feature toggle to the Contact Person section.
- [x] Adding the Person Attributes for the other Contact Person fields (_see Given Name implementation for details_).

File: src/patient-registration/patient-registration.component.tsx [hunk 15]
```diff
@@ -128,11 +132,29 @@ export const PatientRegistration: React.FC = () => {
     };
 
     if (values.contactPersonGivenName) {
-      const contactPersonAttributes = {
-        attributeType: uuidContactPersonGivenName,
-        value: values.contactPersonGivenName,
-      };
-      patient.person.attributes.push(contactPersonAttributes);
+      const contactPersonAttributes = [
```
Comment (nickjhill14) [inline src/patient-registration/patient-registration.component.tsx:131]: We should ideally refactor this before we merge this into `develop` and `master`.

**Review by nickjhill14 — CHANGES_REQUESTED** [review]
Final changes/checks:
- [x] Check the form submits correctly.

**Comment by nickjhill14** [comment]
This ticket is currently blocked because there is currently no Person Attributes for Contact Person. These can be created manually [here](https://openmrs-spa.org/openmrs/admin/person/personAttributeType.list) however, there would be a new UUID generated each time in each instance of OpenMRS. This would mean that the source code would also need to be manually changed to ensure that the correct UUIDs are associated with the Contact Person attributes. Therefore, until a way of dynamically creating Person Attributes (see [MF-282](https://issues.openmrs.org/browse/MF-282)) is achieved. This feature will remain as a branch on `develop`. It's [ticket](https://issues.openmrs.org/browse/MF-216) on JIRA will be edited to reflect this.
